internal/cephfs/snapshot.go

```diff
@@ -0,0 +1,272 @@
+/*
+Copyright 2020 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cephfs
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/ceph/ceph-csi/internal/util"
+
+	"github.com/golang/protobuf/ptypes/timestamp"
+	"k8s.io/klog"
+)
+
+func snapshotNotFoundErrorString(snapName string) string {
+	return fmt.Sprintf("Error ENOENT: snapshot '%s' doesnot exist", snapName)
+}
+
+func snapshotAlreadyProtectedErrorString(snapName string) string {
+	return fmt.Sprintf("Error EEXIST: snapshot '%s' is already protected", snapName)
+}
+
+func snapshotNotProtectedErrorString(snapName string) string {
+	return fmt.Sprintf("Error EEXIST: snapshot '%s' is not protected", snapName)
+}
+
+func createSnapshot(ctx context.Context, volOptions *volumeOptions, cr *util.Credentials, volID volumeID) error {
+	args := []string{
+		"fs",
+		"subvolume",
+		"snapshot",
+		"create",
+		volOptions.FsName,
+		string(volID),
+		volOptions.SnapshotName,
+		"--group_name",
+		volOptions.SubvolumeGroup,
+		"-m", volOptions.Monitors,
+		"-c", util.CephConfigPath,
+		"-n", cephEntityClientPrefix + cr.ID,
+		"--keyfile=" + cr.KeyFile,
+	}
+
+	err := execCommandErr(
+		ctx,
+		"ceph",
+		args[:]...)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to create subvolume snapshot %s %s(%s) in fs %s"), volOptions.SnapshotName, string(volID), err, volOptions.FsName)
+		return err
+	}
+	return nil
+}
+
+func deleteSnapshot(ctx context.Context, volOptions *volumeOptions, cr *util.Credentials, volID volumeID) error {
+	args := []string{
+		"fs",
+		"subvolume",
+		"snapshot",
+		"rm",
+		volOptions.FsName,
+		string(volID),
+		volOptions.SnapshotName,
+		"--group_name",
+		volOptions.SubvolumeGroup,
+		"-m", volOptions.Monitors,
+		"-c", util.CephConfigPath,
+		"-n", cephEntityClientPrefix + cr.ID,
+		"--keyfile=" + cr.KeyFile,
+		"--force",
+	}
+
+	err := execCommandErr(
+		ctx,
+		"ceph",
+		args[:]...)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to delete subvolume snapshot %s %s(%s) in fs %s"), volOptions.SnapshotName, string(volID), err, volOptions.FsName)
+		return err
+	}
+	return nil
+}
+
+type snapshotInfo struct {
+	CreatedAt        string `json:"created_at"`
+	CreationTime     *timestamp.Timestamp
+	ID               string
```

Comment:
ID, Name seems to be overloaded into this info structure. Where needed a snapshotID should be passed along as needed.